### PR TITLE
Skip display mode setting if already set

### PR DESCRIPTION
**Pull request by zvova7890**
That mostly taken from wine's code. Before setting new display mode, we are checking if our current display mode isn't equal new required mode, if new mode is same as current, do not call ChangeDisplaySettingsExW. Excludes display flickering if the application several times setup same display mode

### CODE REVIEW

**Comment by doitsujin** [comment]
I'm fine with the functionality change, but this introduces a non-trivial amount of code-duplication. @Joshua-Ashton what exactly was the reason again for having all the DXGI monitor code duplicated in D3D9, save struct conversion?

Also, there's an unrelated commit that's already being tracked in #1436.

File: src/dxgi/dxgi_monitor.cpp [hunk 1]
```diff
@@ -142,6 +145,26 @@ namespace dxvk {
...
+    if (EnumDisplaySettingsW(monInfo.szDevice, ENUM_CURRENT_SETTINGS, &currentMode)) {
+      if (currentMode.dmPelsWidth == devMode.dmPelsWidth &&
+        currentMode.dmPelsHeight == devMode.dmPelsHeight &&
+        currentMode.dmBitsPerPel == devMode.dmBitsPerPel &&
+        (currentMode.dmDisplayFrequency == devMode.dmDisplayFrequency ||
+        !(devMode.dmFields & DM_DISPLAYFREQUENCY)) &&
+        (currentMode.dmDisplayFlags == devMode.dmDisplayFlags ||
+        !(devMode.dmFields & DM_DISPLAYFLAGS)))
+      {
+        Logger::info(str::format("DXGI: Skipping redundant mode setting call: ",
```
Comment (doitsujin) [inline src/dxgi/dxgi_monitor.cpp]: I highly doubt we need a log message to inform the user that nothing happened.

File: src/dxgi/dxgi_monitor.cpp [hunk 2]
```diff
@@ -142,6 +145,26 @@ namespace dxvk {
...
     
+    /* Compare new mode with current and change if necessary */
+    if (EnumDisplaySettingsW(monInfo.szDevice, ENUM_CURRENT_SETTINGS, &currentMode)) {
+      if (currentMode.dmPelsWidth == devMode.dmPelsWidth &&
+        currentMode.dmPelsHeight == devMode.dmPelsHeight &&
+        currentMode.dmBitsPerPel == devMode.dmBitsPerPel &&
+        (currentMode.dmDisplayFrequency == devMode.dmDisplayFrequency ||
+        !(devMode.dmFields & DM_DISPLAYFREQUENCY)) &&
+        (currentMode.dmDisplayFlags == devMode.dmDisplayFlags ||
+        !(devMode.dmFields & DM_DISPLAYFLAGS)))
```
Comment (doitsujin) [inline src/dxgi/dxgi_monitor.cpp]: `(devMode.dmFields & DM_DISPLAYFLAGS)` is always `false`, so this check should probably be removed.

**Comment by Joshua-Ashton** [comment]
The code duplication really only exists because I wasn't exactly sure what I needed when I implemented, and at that time the D3D9 frontend was still a fork so I couldn't just move stuff around willy nilly there. 😛

Ideally this is the sort of thing that the wsi abstraction in the native port solves...

File: src/dxgi/dxgi_monitor.cpp (hunk 1 above)
Comment (zvova7890) [reply src/dxgi/dxgi_monitor.cpp]: Agree, but if this will have some problems, log message will help with debugging. If it's too much, I will remove

**Comment by doitsujin** [comment]
Well I guess factoring out the monitor code is an option. I kind of don't want to deal with the native stuff right now.

**Comment by zvova7890** [comment]
How about merge this, and then thinking about code unify? :) Code duplication is already present and will not get worse after merge

**Comment by doitsujin** [comment]
> and then thinking about code unify?

We all know that that's not going to happen and instead we'll end up with bit rot.

I've been busy working on vkd3d lately so not much time to work on this.

**Comment by doitsujin** [comment]
Superseded by #1494. This reworks the monitor stuff to remove code duplication, and includes the feature discussed in this PR based on your work (I added you as co-author to give credit, not sure if there's any better way to do this).

**Comment by zvova7890** [comment]
Thank you, i had test you branch, looks fine

**Pull request closed without merging**